DO NOT MERGE: Pods with the DisruptionTarget condition should be treated as terminating by the endpoints controller

smarterclayton opened the pull request:
Test a possible fix to the issue described in https://github.com/kubernetes/kubernetes/issues/116965#issuecomment-2195721381.

After discussion with @aojea and @mimowo and @alculquicondor, this has been updated to leverage (and slightly expand) the scope of the DisruptionTarget condition already used by JobController.

When the Kubelet expects to terminate a pod abnormally (i.e. not as a result of the pod exiting on its own or a deletionTimestamp), we use that info to always set a DisruptionTarget condition (and to override DisruptionTargets set by others).  We also correct a bug in Kubelet state transitions on termination that could delay sending condition changes unpredictably - the condition is now returned immediately after termination begins.  Since we now have a signal that the Kubelet intends to terminate a pod unexpectedly, the endpoints controller can safely use that condition to remove that pod from rotation, thus allowing network proxies the maximum amount of time to react to upcoming disruption.

TODO:
- [ ] Prove that this has no compatibility impact to existing workloads
- [ ] Identify a principle to guide addition of extra conditions if DisruptionTarget alone cannot cover the scenario
- [ ] Follow up to ensure probers continue running until the pods containers are stopped as discussed in #125558
- [ ] Separate the other commits into their own PRs - they are to aid in debugging or noticed while fixing
- [ ] Add better e2e suites around pod termination and ensure DisruptionTarget is set eagerly (at termination time) as well as simulate a graceful node shutdown and restart to ensure the condition is cleared
- [x] Add a reverse feature gate (on by default) to allow disabling the endpoints behavior if we find it breaks unexpectedly
- [ ] Look at the DisruptionTarget GC mechanism and modify it to account for the Kubelet now taking stronger ownership over reason: TerminatedByKubelet (i.e. we should let the kubelet own that state).
- [ ] Open a KEP PR that clarifies the behavior and expectations (to both DisruptionTarget and GracefulShutdown)

/kind bug

Fixes #116965, #124648

```release-note
Set the DisruptionTarget condition as soon as pods begin terminating.  Update the endpoints controller and endpointslice controller to consider a pod with DisruptionTarget to be "terminating", consistent with deletionTimestamp being set.  Add a feature gate `PodDisruptionConditionSignalsTerminating` that defaults to on but can be disabled if the feature causes impacts. Ensure the Kubelet removes the condition after a graceful node shutdown and reboot is completed.

This gives load balancers and proxies exposing the pods the maximum amount of time to remove pods from rotation before the Kubelet terminates the pod.
```
```docs
```

smarterclayton committed:
trigger shutdown with USR1

smarterclayton committed:
hack: Allow local-up-cluster to test graceful node shutdown

Propagate the config flag to the kubelet config.

smarterclayton committed:
kubelet: Update the API description of node graceful shutdown

Clearly describe the behavior when grace period default value is
set.

smarterclayton committed:
kubelet: Log when graceful shutdown is not enabled to aid debugging

During kubelet startup, print a line which indicates whether graceful
deletion is configured to aid admins and debuggers in verifying that
the code is actually running.

smarterclayton committed:
hack: local-up-cluster should not wait for coredns without DNS

We are assuming that coredns is always set up, which is not true
in all possible flags.

This commit does not fix the issue that the implementation may
not be coreDNS and that other DNS_ADDON values may need different
implementations.

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -2176,7 +2176,7 @@ func (kl *Kubelet) SyncTerminatedPod(ctx context.Context, pod *v1.Pod, podStatus
 	// generate the final status of the pod
 	// TODO: should we simply fold this into TerminatePod? that would give a single pod update
-	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true)
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true, true)
```
ok, this is clear, this has already terminated

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -2116,7 +2116,7 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 	// information about the container end states (including exit codes) - when
 	// SyncTerminatedPod is called the containers may already be removed.
-	apiPodStatus = kl.generateAPIPodStatus(pod, podStatus, true)
+	apiPodStatus = kl.generateAPIPodStatus(pod, podStatus, true, true)
```
ok, this is also clear, nothing running here

aojea commented on pkg/kubelet/kubelet.go:
```diff
@@ -2039,7 +2039,7 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 	defer klog.V(4).InfoS("SyncTerminatingPod exit", "pod", klog.KObj(pod), "podUID", pod.UID)
 
-	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, false)
+	apiPodStatus := kl.generateAPIPodStatus(pod, podStatus, true, false)
```
this seems to be the important one

aojea replied:
readiness probes are removed on line 2068, before the pod is killed
```
	if err := kl.killPod(ctx, pod, p, gracePeriod); err != nil {
		kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
		// there was an error killing the pod, so we return that error directly
		utilruntime.HandleError(err)
		return err
	}
```

smarterclayton committed:
endpoints: A pod with DisruptionTarget condition is considered to be terminating

The DisruptionTarget condition is set on pods when either an API client
or Kubelet believes the pod will be interrupted in the near future. Since
the condition is set in advance of the disruption (on the Kubelet as soon
as termination starts, by the scheduler before the pod is deleted), the
endpoints controller can eagerly begin propagating the pod as being
removed from endpoints or set terminating: true in an endpoint slice.

This allows kubelet-initiated disruption to be equivalent to API initiated
disruption (via deletionTimestamp) for the purposes of removing a pod from
rotation as soon as possible, without altering the pod readiness state.

A new feature gate is added to support this behavior in the on state to
allow GracefulNodeShutdown to correctly mitigate issues end users face
with shutdown. The gate defaults on and is intended to allow opt out for
consumers if issues are encountered.

aojea commented:
@smarterclayton I thought we were going to also enhance the endpoint slice controller to set the terminating condition 

https://github.com/kubernetes/kubernetes/blob/d902351c991a68fa76de9935a485afeb1f780c11/staging/src/k8s.io/endpointslice/utils.go#L37-L58

This will allow, from the network perspective of the cluster, to do the right things, workloads we know are going to be terminated because of reasons, can indicate they should only be used if there are no other workload ready

smarterclayton commented:
>  thought we were going to also enhance the endpoint slice controller to set the terminating condition based on the Pod condition

https://github.com/kubernetes/kubernetes/pull/125774/files#diff-fc23208966763e11ddaea01723e972e3862da2f5a25d2f4c98e00b199655f23aR44

smarterclayton committed:
kubelet: Set a condition when pods are terminating on the Kubelet

The DisruptionTarget condition is used to signal that the pod was
shutdown abnormally for reasons outside the pod's control in order
to allow job controllers to restart the pod without considering it
a full failure. Some causes of abnormal pod termination are not
covered today which are important to signal that pods should be
removed from endpoints. The PendingTermination condition is set
whenever an abnormal termination of a pod is begun on the Kubelet
and will be used by endpoints controllers to preemptively remove
that pod from serving to avoid traffic disruption.

smarterclayton committed:
kubelet: Propagate the DisruptionTarget condition aggressively

The Kubelet wants the DisruptionTarget condition to show up in
the API as quickly as possible to allow endpoints to react to
future disruption.

smarterclayton committed:
endpoints: A pod that is PendingTermination is considered to be terminating

Extend the previous support for pod disruption condition signalling
to use the broader PendingTermination condition and catch other
Kubelet conditions. We continue to leverage DisruptionTarget because
it gives us additional advance notice for disruption.

smarterclayton committed:
kubelet: Clearly signal abnormal pod termination

The pod worker owns the lifecycle of the pod, and is responsible
for knowing the difference between a clean termination (due to
the behavior of the pod's lifecycle) and an abnormal termination
(due to events on the Kubelet). Refactor the pod killNow logic
into a clear method on the pod worker indicating an abnormal
termination, then have the pod worker pass that along to
SyncTerminatingPod.

As a result, the SyncTerminatingPod method now knows whether the
pod's lifecycle ended or was interrupted, which can be used to
signal abnormal disruption of the pod.

As a consequence of this change, abnormal termination is limited
to failing the pod, setting a reason and message, and a set of
optional conditions to prevent misuse.

This also forces the phase change caused by abnormal termination
to be handled explicitly in generateAPIPodStatus and other phase
change logic is now invoked.
